File: src/main/java/com/flexpoker/table/query/repository/impl/RedisCardsUsedInHandRepository.java
package com.flexpoker.table.query.repository.impl;

import java.util.Map;
import java.util.UUID;
import java.util.function.Function;
import java.util.stream.Collectors;

import javax.inject.Inject;

import org.springframework.context.annotation.Profile;
import org.springframework.data.redis.core.RedisTemplate;
import org.springframework.stereotype.Repository;

import com.flexpoker.config.ProfileNames;
import com.flexpoker.model.card.FlopCards;
import com.flexpoker.model.card.PocketCards;
import com.flexpoker.model.card.RiverCard;
import com.flexpoker.model.card.TurnCard;
import com.flexpoker.table.query.repository.CardsUsedInHandRepository;

@Profile({ ProfileNames.REDIS, ProfileNames.TABLE_QUERY_REDIS })
@Repository
public class RedisCardsUsedInHandRepository implements CardsUsedInHandRepository {

    private static final String FLOP_CARDS_NAMESPACE = "flop-cards:";

    private static final String TURN_CARD_NAMESPACE = "turn-card:";

    private static final String RIVER_CARD_NAMESPACE = "river-card:";

    private static final String HAND_POCKET_CARDS_NAMESPACE = "hand-pocket-cards:";

    private final RedisTemplate<String, String> stringRedisTemplate;

    private final RedisTemplate<String, FlopCards> redisTemplateFlopCards;

    private final RedisTemplate<String, TurnCard> redisTemplateTurnCard;

    private final RedisTemplate<String, RiverCard> redisTemplateRiverCard;

    private final RedisTemplate<String, Map<UUID, PocketCards>> redisTemplatePocketCards;

    @Inject
    public RedisCardsUsedInHandRepository(
            RedisTemplate<String, String> stringRedisTemplate,
            RedisTemplate<String, FlopCards> redisTemplateFlopCards,
            RedisTemplate<String, TurnCard> redisTemplateTurnCard,
            RedisTemplate<String, RiverCard> redisTemplateRiverCard,
            RedisTemplate<String, Map<UUID, PocketCards>> redisTemplatePocketCards) {
        this.stringRedisTemplate = stringRedisTemplate;
        this.redisTemplateFlopCards = redisTemplateFlopCards;
        this.redisTemplateTurnCard = redisTemplateTurnCard;
        this.redisTemplateRiverCard = redisTemplateRiverCard;
        this.redisTemplatePocketCards = redisTemplatePocketCards;
    }

    @Override
    public void saveFlopCards(UUID handId, FlopCards flopCards) {
        redisTemplateFlopCards.opsForValue().set(FLOP_CARDS_NAMESPACE + handId, flopCards);
    }

    @Override
    public void saveTurnCard(UUID handId, TurnCard turnCard) {
        redisTemplateTurnCard.opsForValue().set(TURN_CARD_NAMESPACE + handId, turnCard);
    }

    @Override
    public void saveRiverCard(UUID handId, RiverCard riverCard) {
        redisTemplateRiverCard.opsForValue().set(RIVER_CARD_NAMESPACE + handId, riverCard);
    }

    @Override
    public void savePocketCards(UUID handId, Map<UUID, PocketCards> playerToPocketCardsMap) {
        redisTemplatePocketCards.opsForHash().putAll(HAND_POCKET_CARDS_NAMESPACE + handId, playerToPocketCardsMap);
    }

    @Override
    public FlopCards fetchFlopCards(UUID handId) {
        return redisTemplateFlopCards.opsForValue().get(FLOP_CARDS_NAMESPACE + handId);
    }

    @Override
    public TurnCard fetchTurnCard(UUID handId) {
        return redisTemplateTurnCard.opsForValue().get(TURN_CARD_NAMESPACE + handId);
    }

    @Override
    public RiverCard fetchRiverCard(UUID handId) {
        return redisTemplateRiverCard.opsForValue().get(RIVER_CARD_NAMESPACE + handId);
    }

    @Override
    public PocketCards fetchPocketCards(UUID handId, UUID playerId) {
        return fetchPocketCardsForHand(handId).get(playerId);
    }

    @Override
    public Map<UUID, PocketCards> fetchAllPocketCardsForUser(UUID playerId) {
        return stringRedisTemplate.keys(HAND_POCKET_CARDS_NAMESPACE + "*")
                .stream()
                .map(keyString -> UUID.fromString(keyString.substring(HAND_POCKET_CARDS_NAMESPACE.length())))
                .collect(Collectors.toMap(
                        Function.identity(),
                        handId -> fetchPocketCardsForHand(handId)))
                .entrySet()
                .stream()
                .filter(x -> x.getValue().containsKey(playerId))
                .collect(Collectors.toMap(Map.Entry::getKey, x -> x.getValue().get(playerId)));
    }

    private Map<UUID, PocketCards> fetchPocketCardsForHand(UUID handId) {
        return redisTemplatePocketCards.opsForHash().entries(HAND_POCKET_CARDS_NAMESPACE + handId)
                .entrySet()
                .stream()
                .collect(Collectors.toMap(
                        x -> ((UUID) x.getKey()),
                        x -> (PocketCards) x.getValue()));
    }

}

File: src/main/kotlin/com/flexpoker/table/query/repository/impl/RedisCardsUsedInHandRepository.kt
package com.flexpoker.table.query.repository.impl

import com.flexpoker.config.ProfileNames
import com.flexpoker.table.command.FlopCards
import com.flexpoker.table.command.PocketCards
import com.flexpoker.table.command.RiverCard
import com.flexpoker.table.command.TurnCard
import com.flexpoker.table.query.repository.CardsUsedInHandRepository
import org.springframework.context.annotation.Profile
import org.springframework.data.redis.core.RedisTemplate
import org.springframework.data.redis.core.ScanOptions
import org.springframework.stereotype.Repository
import java.util.UUID
import javax.inject.Inject

@Profile(ProfileNames.REDIS, ProfileNames.TABLE_QUERY_REDIS)
@Repository
class RedisCardsUsedInHandRepository @Inject constructor(
    private val stringRedisTemplate: RedisTemplate<String, String>,
    private val redisTemplateFlopCards: RedisTemplate<String, FlopCards>,
    private val redisTemplateTurnCard: RedisTemplate<String, TurnCard>,
    private val redisTemplateRiverCard: RedisTemplate<String, RiverCard>,
    private val redisTemplatePocketCards: RedisTemplate<String, Map<UUID, PocketCards>>,
) : CardsUsedInHandRepository {

    companion object {
        private const val FLOP_CARDS_NAMESPACE = "flop-cards"
        private const val TURN_CARD_NAMESPACE = "turn-card"
        private const val RIVER_CARD_NAMESPACE = "river-card"
        private const val HAND_POCKET_CARDS_NAMESPACE = "hand-pocket-cards"

        private val HAND_POCKET_CARDS_SCAN_OPTIONS = ScanOptions.scanOptions()
            .match("$HAND_POCKET_CARDS_NAMESPACE:*")
            .count(100)
            .build()
    }

    private fun flopCardsRedisKey(handId: UUID) = "$FLOP_CARDS_NAMESPACE:$handId"
    private fun turnCardRedisKey(handId: UUID) = "$TURN_CARD_NAMESPACE:$handId"
    private fun riverCardRedisKey(handId: UUID) = "$RIVER_CARD_NAMESPACE:$handId"
    private fun handPocketCardsRedisKey(handId: UUID) = "$HAND_POCKET_CARDS_NAMESPACE:$handId"

    override fun saveFlopCards(handId: UUID, flopCards: FlopCards) {
        redisTemplateFlopCards.opsForValue()[flopCardsRedisKey(handId)] = flopCards
    }

    override fun saveTurnCard(handId: UUID, turnCard: TurnCard) {
        redisTemplateTurnCard.opsForValue()[turnCardRedisKey(handId)] = turnCard
    }

    override fun saveRiverCard(handId: UUID, riverCard: RiverCard) {
        redisTemplateRiverCard.opsForValue()[riverCardRedisKey(handId)] = riverCard
    }

    override fun savePocketCards(handId: UUID, playerToPocketCardsMap: Map<UUID, PocketCards>) {
        redisTemplatePocketCards.opsForHash<Any, Any>()
            .putAll(handPocketCardsRedisKey(handId), playerToPocketCardsMap)
    }

    override fun fetchFlopCards(handId: UUID): FlopCards? {
        return redisTemplateFlopCards.opsForValue()[flopCardsRedisKey(handId)]
    }

    override fun fetchTurnCard(handId: UUID): TurnCard? {
        return redisTemplateTurnCard.opsForValue()[turnCardRedisKey(handId)]
    }

    override fun fetchRiverCard(handId: UUID): RiverCard? {
        return redisTemplateRiverCard.opsForValue()[riverCardRedisKey(handId)]
    }

    override fun fetchPocketCards(handId: UUID, playerId: UUID): PocketCards? {
        return fetchPocketCardsForHand(handId)[playerId]
    }

    override fun fetchAllPocketCardsForUser(playerId: UUID): Map<UUID, PocketCards> {
        val cursor = stringRedisTemplate.scan(HAND_POCKET_CARDS_SCAN_OPTIONS)

        val keys = mutableListOf<String>()
        while (cursor.hasNext()) {
            keys.add(cursor.next())
        }

        return keys
            .associateWith { redisTemplatePocketCards.opsForHash<UUID, PocketCards>().get(it, playerId) }
            .filter { it.value != null }
            .map { (key, value) -> UUID.fromString(key.substring(HAND_POCKET_CARDS_NAMESPACE.length.inc())) to value!! }
            .toMap()
    }

    override fun removeHand(handId: UUID) {
        redisTemplateFlopCards.delete(flopCardsRedisKey(handId))
        redisTemplateTurnCard.delete(turnCardRedisKey(handId))
        redisTemplateRiverCard.delete(riverCardRedisKey(handId))
        redisTemplatePocketCards.delete(handPocketCardsRedisKey(handId))
    }

    private fun fetchPocketCardsForHand(handId: UUID): Map<UUID, PocketCards> {
        return redisTemplatePocketCards.opsForHash<UUID, PocketCards>()
            .entries(handPocketCardsRedisKey(handId))
    }

}